src/ota_metadata/file_table/_table.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2022 TIER IV, INC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+
+import os
+import shutil
+import stat
+from pathlib import Path
+from typing import ClassVar, Literal, Optional
+
+from pydantic import BaseModel, SkipValidation
+from simple_sqlite3_orm import ConstrainRepr, TableSpec, TypeAffinityRepr
+from typing_extensions import Annotated
+
+from ota_metadata.file_table._types import EntryAttrsType
+from otaclient_common.typing import StrOrPath
+
+CANONICAL_ROOT = "/"
+
+PrepareMethod = Literal["move", "hardlink", "copy"]
+
+
+class FileTableBase(BaseModel):
+    schema_ver: ClassVar[Literal[1]] = 1
+
+    path: Annotated[
+        str,
+        TypeAffinityRepr(str),
+        ConstrainRepr("PRIMARY KEY"),
+        SkipValidation,
+    ]
+
+    entry_attrs: Annotated[
+        EntryAttrsType,
+        TypeAffinityRepr(bytes),
+        ConstrainRepr("NOT NULL"),
+    ]
+    """msgpacked basic attrs for this file entry.
+
+    Ref: https://www.kernel.org/doc/html/latest/filesystems/ext4/inodes.html
+
+    including the following fields from inode table:
+    1. mode bits
+    2. uid
+    3. gid
+    4. inode(when the file is hardlinked)
+    5. xattrs
+
+    See file_table._types for more details.
+    """
+
+    def set_xattr(self, _target: StrOrPath) -> None:
+        """Set the xattr of self onto the <_target>.
+
+        NOTE: this method always don't follow symlink.
+        """
+        if xattrs := self.entry_attrs.xattrs:
+            for k, v in xattrs.items():
+                os.setxattr(
+                    path=_target,
+                    attribute=k,
+                    value=v.encode(),
+                    follow_symlinks=False,
+                )
+
+    def set_perm(self, _target: StrOrPath) -> None:
+        """Set the mode,uid,gid of self onto the <_target>."""
+        entry_attrs = self.entry_attrs
+        # NOTE(20241213): chown will reset the sticky bit of the file!!!
+        #   Remember to always put chown before chmod !!!
```

Comment:
Great comment.

Reply:
This pitfall is caught by the test, test indeed is very important. :test_tube: